src/Database/Oracle/Simple/Internal.hs

```diff
@@ -818,44 +873,125 @@ stmtRelease
   -> IO ()
 stmtRelease = throwOracleError <=< dpiStmt_release
 
-data DPIData = DPIData
+-- | Used to write values to or read values from a column.
+data DPIData a = DPIData
   { dataIsNull :: CInt
-  , dataValue :: DPIDataBuffer
+  -- ^ If reading, a null value was read. If writing, writes a null value.
+  , dataValue :: a
+  -- ^ The value that was read/will be written, of type 'ReadBuffer' or 'WriteBuffer'.
   }
   deriving stock (Generic, Show, Eq)
   deriving anyclass (GStorable)
 
-newtype DPIDataBuffer = DPIDataBuffer (Ptr DPIDataBuffer)
+-- | An opaque pointer type for the @dpiDataBuffer@ union that we read from.
+-- We cannot write to this in a way that ODPIC could use.
+-- For poking purposes, use 'WriteBuffer'.
+newtype ReadBuffer = ReadBuffer (Ptr ReadBuffer)
   deriving (Show, Eq)
   deriving newtype (Storable)
 
+-- | @dpiDataBuffer@ union that we can write to.
+-- We cannot read from this without a hint as to what type of data it contains.
+data WriteBuffer
+  = AsBoolean CBool
+  | AsInt64 Int64
+  | AsUInt64 Word64
+  | AsFloat Float
+  | AsDouble Double
+  | AsString CString
+  | AsBytes DPIBytes
+  | AsTimestamp DPITimestamp
+  | AsNull (Ptr ())
+  deriving (Show, Eq, Generic)
+
+instance Storable WriteBuffer where
+  sizeOf _ = sizeOf (undefined :: DPITimestamp) -- size of largest element
+  alignment _ = 8
+  peek = error "WriteBuffer says: don't peek in here!" -- handle this better!
+
+  poke ptr (AsInt64 intVal) = poke (castPtr ptr) intVal
+  poke ptr (AsUInt64 word64Val) = poke (castPtr ptr) word64Val
+  poke ptr (AsFloat floatVal) = poke (castPtr ptr) floatVal
+  poke ptr (AsDouble doubleVal) = poke (castPtr ptr) doubleVal
+  poke ptr (AsString cStringVal) = poke (castPtr ptr) cStringVal
+  poke ptr (AsBytes dpiBytesVal) = poke (castPtr ptr) dpiBytesVal
+  poke ptr (AsTimestamp dpiTimeStampVal) = poke (castPtr ptr) dpiTimeStampVal
+  poke ptr (AsNull nullVal) = poke (castPtr ptr) nullVal
```

Comment:
Do we need `AsNull (Ptr ())` or could we just keep it `AsNull`, and then reference `nullPtr`.

So for the above ^

```haskell
  poke ptr (AsNull nullVal) = poke (castPtr ptr) nullVal
```

would become

```haskell
  poke ptr AsNull = poke (castPtr ptr) nullPtr
  ```